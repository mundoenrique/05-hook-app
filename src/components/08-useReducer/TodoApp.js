import { useEffect, useReducer } from 'react';

import { todoReducer } from '../../helpers/toolHelpers';
import TodoList from './TodoList';
import TodoAdd from './TodoAdd';

const initReducer = () => {
	return JSON.parse(localStorage.getItem('todos')) || [];
};

export default function TodoApp() {
	const [todos, dispatch] = useReducer(todoReducer, [], initReducer);

	useEffect(() => {
		localStorage.setItem('todos', JSON.stringify(todos));
	}, [todos]);

	const handleDelete = (todoId) => {
		const action = {
			type: 'delete',
			payload: todoId,
		};

		dispatch(action);
	};

	const handleToggle = (todoId) => {
		dispatch({
			type: 'toggle',
			payload: todoId,
		});
	};

	const handleAddTodo = (newTodo) => {
		dispatch({
			type: 'add',
			payload: newTodo,
		});
	};

	return (
		<>
			<h3>TodoApp ({todos.length})</h3>
			<div className="row">
				<div className="col-7">
					<TodoList
						todos={todos}
						handleDelete={handleDelete}
						handleToggle={handleToggle}
					/>
				</div>
				<div className="col-5">
					<TodoAdd handleAddTodo={handleAddTodo} />
				</div>
			</div>
		</>
	);
}
