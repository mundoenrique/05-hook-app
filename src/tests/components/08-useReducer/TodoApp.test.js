import { act } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import TodoApp from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Pruebas sobre <TodoApp />', () => {
	const wrapper = shallow(<TodoApp />);
	Storage.prototype.setItem = jest.fn(() => {});

	test('Debe renderizar <TodoApp />', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('Debe agregar uno TODO', () => {
		const wrapper = mount(<TodoApp />);

		act(() => {
			wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
			wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1]);
		});

		expect(wrapper.find('h3').text().trim()).toBe(
			`TodoApp (${demoTodos.length})`
		);

		expect(localStorage.setItem).toHaveBeenCalledTimes(demoTodos.length);
	});

	test('Debe eliminar un TOdo', () => {
		wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);

		expect(wrapper.find('h3').text().trim()).toBe(`TodoApp (1)`);

		wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);

		expect(wrapper.find('h3').text().trim()).toBe(`TodoApp (0)`);
	});
});
