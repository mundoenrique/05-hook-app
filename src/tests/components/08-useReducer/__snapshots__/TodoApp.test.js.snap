// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <TodoApp /> Debe renderizar <TodoApp /> 1`] = `
<Fragment>
  <h3>
    TodoApp (
    0
    )
  </h3>
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
